File: Algorithms/CSharp/RandomSearch/RandomSearch.cs
using System;
using System.Collections.Generic;
using System.Linq;
using OSOL.Extremum.Cores.DotNet.Optimization;
using OSOL.Extremum.Cores.DotNet.Vectors;
using OSOL.Extremum.Cores.DotNet.Optimization.Nodes;
using OSOL.Extremum.Cores.DotNet.Random;
using OSOL.Extremum.Cores.DotNet.Random.Distributions;

namespace OSOL.Extremum.Algorithms.CSharp
{
    
    using Area = Dictionary<string, Tuple<double, double>>;
    
    public static class RandomSearch
    {
        private static string currentPointName = "currentPoint";
        private static string currentPointEfficiencyName = "currentPointEfficiency";
        private static string radiusParameterName = "r";
        private static GoRN gorn = new GoRN();

        private static RealVector GenerateRandomInSpere(RealVector currentPoint, double radius, Area area)
        {
            RealVector normallyDistributed = gorn.GetNormalVector(area.ToDictionary(kvp => kvp.Key, kvp => Tuple.Create(0.0, 1.0)));
            var r = Math.Sqrt(normallyDistributed.Elements.Values.Select(v => v * v).Sum());
            RealVector generatedPoint = (currentPoint +
                                         (RealVector) (normallyDistributed *
                                                       (gorn.GetContinuousUniform(-1.0, 1.0) / r)));
            return generatedPoint.Constrain(area);
        }

        private class GenerateInitialPointNode : GeneralNode<RealVector, double, RealVector>
        {

            public GenerateInitialPointNode(int nodeId)
            {
                this.NodeId = nodeId;
            }
            
            public override void Initialize(Func<Dictionary<string, double>, double> f, Area area, State<RealVector, double, RealVector> state)
            {
                RealVector initialPoint = gorn.GetContinuousUniformVector(area);
                state.SetParameter(currentPointName, initialPoint);
                state.SetParameter(currentPointEfficiencyName, initialPoint.GetPerformance(f));
            }

            public override void Process(Func<Dictionary<string, double>, double> f, Area area, State<RealVector, double, RealVector> state)
            {
                
            }
        }

        private class SampleNewPointNode_FixedStep : GeneralNode<RealVector, double, RealVector>
        {
            public SampleNewPointNode_FixedStep(int nodeId)
            {
                this.NodeId = nodeId;
            }

            public override void Initialize(Func<Dictionary<string, double>, double> f, Area area, State<RealVector, double, RealVector> state)
            {
                
            }

            public override void Process(Func<Dictionary<string, double>, double> f, Area area, State<RealVector, double, RealVector> state)
            {
                var currentPoint = state.GetParameter<RealVector>(currentPointName);
                var currentPointEfficiency = state.GetParameter<double>(currentPointEfficiencyName);
                var r = state.GetParameter<double>(radiusParameterName);

                var newPoint = GenerateRandomInSpere(currentPoint, r, area);
                var newPointEfficiency = newPoint.GetPerformance(f);

                if (newPointEfficiency < currentPointEfficiency)
                {
                    state.SetParameter(currentPointName, newPoint);
                    state.SetParameter(currentPointEfficiencyName, newPointEfficiency);
                }
            }
        }
        
        private class SetBestNode : GeneralNode<RealVector, double, RealVector>
        {
            public SetBestNode(int nodeId)
            {
                this.NodeId = nodeId;
            }

            public override void Initialize(Func<Dictionary<string, double>, double> f, Area area, State<RealVector, double, RealVector> state)
            {
                
            }

            public override void Process(Func<Dictionary<string, double>, double> f, Area area, State<RealVector, double, RealVector> state)
            {
                state.result = state.GetParameter<RealVector>(currentPointName);
            }
        }

        public static Algorithm<RealVector, double, RealVector> CreateFixedStepRandomSearch(double radius, double maxTime)
        {
            var FixedStep_nodes = new GeneralNode<RealVector, double, RealVector>[]
            {
                new SetParametersNode<RealVector, double, RealVector>(nodeId: 0, parameters: new Dictionary<string, object> {{radiusParameterName, radius}}),
                new GenerateInitialPointNode(nodeId: 1),
                new TerminationViaMaxTime<RealVector, double, RealVector>(nodeId: 2, maxTime: maxTime),
                new SampleNewPointNode_FixedStep(nodeId: 3),
                new SetBestNode(nodeId: 4)
            };
            var FixedStep_transitionMatrix = new[]
            {
                Tuple.Create<int, int?, int>(0, null, 1),
                Tuple.Create<int, int?, int>(1, null, 2),
                Tuple.Create<int, int?, int>(2, 0, 3),
                Tuple.Create<int, int?, int>(2, 1, 4),
                Tuple.Create<int, int?, int>(3, null, 2)
            };
            
            return new Algorithm<RealVector, double, RealVector>(FixedStep_nodes, FixedStep_transitionMatrix);
        }
        
    }
}

File: Algorithms/FSharp/RandomSearch/RandomSearch.fs
namespace OSOL.Extremum.Algorithms.FSharp

open System
open System.Collections.Generic
open System.Linq
open System.Runtime.CompilerServices
open OSOL.Extremum.Cores.DotNet.Optimization
open OSOL.Extremum.Cores.DotNet.Vectors
open OSOL.Extremum.Cores.DotNet.Optimization.Nodes
open OSOL.Extremum.Cores.DotNet.Random
open OSOL.Extremum.Cores.DotNet.Random.Distributions

module RandomSearch = 

    let CurrentPointName = "currentPoint"
    let CurrentPointEfficiencyName = "currentPointEfficiency"
    let RadiusParameterName = "r"
    let GoRN = new GoRN()
        
    let GenerateRandomInSphere (currentPoint: RealVector) (radius: double) (area: Dictionary<string, double * double>) =
        let normallyDistributed = area.ToDictionary ((fun kvp -> kvp.Key), (fun kvp -> (0.0, 1.0)))
                                  |> GoRN.GetNormalVector
                                  |> RealVector.op_Implicit
                                  
        let r = normallyDistributed.Elements.Values
                |> Seq.map (fun x -> x * x)
                |> Seq.sum
        
        ((currentPoint + normallyDistributed * (GoRN.GetContinuousUniform(-1.0, 1.0) / r)).Elements |> RealVector.op_Implicit).Constrain area
        
    type GenerateInitialPointNode = 
        inherit GeneralNode<RealVector, double, RealVector> 
        
        new (nodeId: int) as this = { } then this.NodeId <- nodeId
        
        override this.Initialize(f: Func<Dictionary<string, double>, double>, area: Dictionary<string, double * double>, state: State<RealVector, double, RealVector>) = 
            let initialPoint: RealVector = GoRN.GetContinuousUniformVector(area) |> RealVector.op_Implicit
            state.SetParameter(CurrentPointName, initialPoint)
            state.SetParameter(CurrentPointEfficiencyName, initialPoint.GetPerformance(f))
            ()
        
        override this.Process(f: Func<Dictionary<string, double>, double>, area: Dictionary<string, double * double>, state: State<RealVector, double, RealVector>) = ()
    
    type SampleNewPointNode_FixedStep = 
        inherit GeneralNode<RealVector, double, RealVector>
        
        new (nodeId: int) as this = { } then this.NodeId <- nodeId
        
        override this.Initialize(f: Func<Dictionary<string, double>, double>, area: Dictionary<string, double * double>, state: State<RealVector, double, RealVector>) = ()
    
        override this.Process(f: Func<Dictionary<string, double>, double>, area: Dictionary<string, double * double>, state: State<RealVector, double, RealVector>) =
            let currentPoint = state.GetParameter<RealVector>(CurrentPointName)
            let currentPointEfficiency = state.GetParameter<Double>(CurrentPointEfficiencyName)
            let r = state.GetParameter<double>(RadiusParameterName)
            let newPoint = GenerateRandomInSphere currentPoint r area
            let newPointEfficiency = newPoint.GetPerformance(f)
            if newPointEfficiency < currentPointEfficiency then
                state.SetParameter(CurrentPointName, newPoint)
                state.SetParameter(CurrentPointEfficiencyName, newPointEfficiency)
            ()
    
    type SetBestNode = 
        inherit GeneralNode<RealVector, double, RealVector> 
        
        new (nodeId: int) as this = { } then this.NodeId <- nodeId
        
        override this.Initialize(f: Func<Dictionary<string, double>, double>, area: Dictionary<string, double * double>, state: State<RealVector, double, RealVector>) = ()
        
        override this.Process(f: Func<Dictionary<string, double>, double>, area: Dictionary<string, double * double>, state: State<RealVector, double, RealVector>) =
            state.result <- state.GetParameter<RealVector>(CurrentPointName)
        
    let CreateFixedStepRandomSearch (radius: double) (maxTime: double) =
        let FixedStep_nodes: GeneralNode<RealVector, double, RealVector>[] = 
            [|
                new SetParametersNode<RealVector, double, RealVector>(nodeId = 0, parameters = [|(RadiusParameterName, radius :> obj)|].ToDictionary ((fun kvp -> fst kvp), (fun kvp -> snd kvp)))
                new GenerateInitialPointNode(nodeId = 1)
                new TerminationViaMaxTime<RealVector, double, RealVector>(nodeId = 2, maxTime = maxTime)
                new SampleNewPointNode_FixedStep(nodeId = 3)
                new SetBestNode(nodeId = 4)
            |]
            
        let FixedStep_transitionMatrix: (int * Nullable<int> * int) [] = 
            [|
                (0, new Nullable<int>(), 1)
                (1, new Nullable<int>(), 2)
                (2, new Nullable<int>(0), 3)
                (2, new Nullable<int>(1), 4)
                (3, new Nullable<int>(), 2)
            |]
        
        new Algorithm<RealVector, double, RealVector>(nodes = FixedStep_nodes, transitionMatrix = FixedStep_transitionMatrix)
